### PR TITLE
[3.13] gh-122695: Fix double-free when using `gc.get_referents` with a freed `_asyncio.FutureIter`

### PR DESCRIPTION
**Pull request by ZeroIntensity**
CC @colesbury

* Issue: gh-122695

### CODE REVIEW

**Comment by colesbury** [comment]
I think it's worth including the test case from #122834

**Comment by ZeroIntensity** [comment]
> I think it's worth including the test case from #122834

Oh yeah, forgot to add that here as well. It's there now.

File: Lib/test/test_asyncio/test_futures.py
Comment (vstinner) [inline Lib/test/test_asyncio/test_futures.py:683]: You should force a `gc.collect()` here.

File: Modules/_asynciomodule.c
Comment (vstinner) [inline Modules/_asynciomodule.c]: I don't think that this comment is useful, I suggest to just remove it.

File: Lib/test/test_asyncio/test_futures.py
Comment (vstinner) [inline Lib/test/test_asyncio/test_futures.py]: The collect() must be done once evil list is created, not before.

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by vstinner** [comment]
Changes in the 3.13 branch currently need approvals from two core devs.

**Review by kumaraditya303 — APPROVED** [review]

**Review by colesbury — APPROVED** [review]

**Pull request merged**
